### PR TITLE
Jsc exponent op

### PR DESCRIPTION
**Pull request by juansc**
Hey you guys! I added exponents and removed the '--' and '++' from our syntax, parser, and scanner. 

Quick!

``` ruby
# Exponents are right associative per mathematical convention
2 ** 3 ** 4 # This is 2 ** (3 ** 4)

# A unary negation binds tighter on the right....
2 ** -3 # This is 1/8

# But binds less tightly on the left
-2 ** 4 # This is -(2 ** 4) = -16, not (-2**4) = 16
```

@rachelriv @alexschneider @whusted Please review and merge.

### CODE REVIEW

**Comment by rtoal** [comment]
Good work on determining the desired precedence and associativity, Juan.

However your grammar

```
 Exp5    ::=  prefixop? Exp6
 Exp6    ::=  (Exp7 '**')* Exp7
 Exp7    ::=  Exp8 (('.' Exp8) | ('[' Exp8 ']') | ('(' arglist ')'))*
 Exp8    ::=  boolLit | intLit | floatLit | id | '(' Exp ')' | StringLit | TupLit | SetLit | MapLit | ListLit | Range | Slice | nonelit
```

does not appear to me to pick up `2 ** -3`  because after the `**` you can only have an Exp7, which cannot begin with a prefixop. Or can it?

**Comment by whusted** [comment]
@rtoal is right, we have held off on exponents for now. What does everyone else think? I'm fine with having them in, we just have to make sure to change the macro to match it.

**Comment by rtoal** [comment]
I have written a prototype grammar that handles exponentiation and unary negation properly on the Keck Lab board. I've explained it to Zane and Alex. Alex gets it. Not sure if Zane does. Before it gets erased, go take a peek. If it's gone, try to come up with it yourself. :)

**Comment by juansc** [comment]
@rtoal Can you or @alexschneider take a picture?

@rtoal I have implemented the following change right now. Does this implementation make sense?

``` python
2 ** -4 ** -3 # This is 2 ** ( -( 4 ** -3 ) )
```

In this way the rightmost unary negation binds to the three, but the unary negation on the four applies to what's to the right of it.

**Comment by juansc** [comment]
I don't mind leaving out exponents, but I don't think the changes to our macrosyntax and our scanner and parser are so great we should leave them out.

**Comment by juansc** [comment]
Here is a proposed modified macrosyntax

```
 Exp5    ::=  prefixop? Exp6
 Exp6    ::=  Exp7 | ( Exp7 '**' (Exp5 '**')* Exp5)
 Exp7    ::=  Exp8 (('.' Exp8) | ('[' Exp8 ']') | ('(' arglist ')'))*
```

I'm not sure this conserves the right-associativity of the exponents though.

**Comment by alexschneider** [comment]
@juansc do you mind merging master onto your branch and resolving any conflicts?

Also, I think that the macrosyntax would be better and clearer this way:

```
Exp6    ::=  Exp7 ('**' Exp5)?
```

**Comment by juansc** [comment]
@alexschneider I'm not sure if that macrosyntax implies that exponentiation is left-associative, rather than right-associative.

**Comment by alexschneider** [comment]
@juansc: I'm not confident that the macrosyntax implies anything about associtivity - I thought that's all done within the parser itself. Check the pic for what @rtoal mentioned

I also changed it slightly since you last saw it, so check the updated comment.

**Comment by rachelriv** [comment]
We can check out Python's grammar for reference.

In Python, the power operator binds more tightly than unary operators on its left; it binds less tightly than unary operators on its right. 

```
power ::=  primary ["**" u_expr]
```

**Comment by rachelriv** [comment]
@juansc It looks like some coffeelint things are failing.

You can test it by running `coffeelint .` from the root of this directory to see why it is complaining.

**Comment by juansc** [comment]
@rachelriv Thanks I will do it right now and let you know if I have questions.

**Comment by alexschneider** [comment]
@rachelriv @juansc, if you use [SublimeLinter](http://www.sublimelinter.com/en/latest/) it will tell you about these issues as you're writing the code.

**Comment by juansc** [comment]
The only thing that fails coffeelint now is the coffee-coverage. I don't touch the stuff. Everything else should work.

**Comment by alexschneider** [comment]
Also, I would feel more comfortable if there was a separate test for the `**` operator, but if @rachelriv or @whusted  is comfortable with merging this as is, then LGTM.

**Comment by whusted** [comment]
@juansc I'd like to see a quick exponent test as a sanity check before merging

**Comment by juansc** [comment]
There are 5 tests. In the seventh teascript program I had added 

```
3 ** 4 + 5
-2 ** 4 + 5
3 ** 4 ** 5
2 ** 1 ** -3 + 1
2 ** -3 ** -4
```

these get parsed into 

```
(+ (** 3 4) 5)
(+ (- (** 2 4)) 5)
(** 3 (** 4 5))
(+ (** 2 (** 1 (- 3))) 1)
(** 2 (- (** 3 (- 4))))
```

**Comment by alexschneider** [comment]
@juansc: I think what Willy meant is that it should be in a separate teascript program with a separate test (separate describe, it, etc)

**Comment by juansc** [comment]
Can I just create program14.tea and copy over those tests?

**Comment by alexschneider** [comment]
I've already created program13.tea in one of my merges, I'd prefer if you created program14.tea.

**Comment by juansc** [comment]
That's what I said program14.tea

**Comment by whusted** [comment]
That's what I meant, thanks for clarifying, Alex. @juansc it's more to match the format of how we're testing and keep everything consistent

**Comment by alexschneider** [comment]
@juansc I've got photographic evidence!

**Comment by juansc** [comment]
Cool I just added it.

**Comment by juansc** [comment]
@alexschneider It's obviously been doctored.

File: README.md [hunk 1]
```diff
@@ -68,7 +68,7 @@ Exp2    ::=  Exp3 (relop Exp3)?
 Exp3    ::=  Exp4 (addop Exp4)*
 Exp4    ::=  Exp5 (mulop Exp5)*
 Exp5    ::=  prefixop? Exp6
-Exp6    ::=  (Exp7 unary?) | (unary Exp7)
+Exp6    ::=  (Exp7 '**')* Exp7
```
Comment (alexschneider) [inline README.md]: We haven't solved our problem with operator precedence - this should be `Exp6 ::= Exp7 ('**' Exp5)?`

File: test/example/output_ASTs.coffee [hunk 2]
```diff
@@ -87,4 +85,11 @@ class MyASTs
       (Invoke e ('arg1', 'arg2'))))))
       (Invoke (Invoke f (a, b)) (c))))"
 
+  @program14_AST_string = "(Program (Block
```
Comment (alexschneider) [inline test/example/output_ASTs.coffee]: I hate to be a stickler, but we don't need double quotes here. Single quotes only unless necessary, to keep consistency across program.

**Comment by rachelriv** [comment]
@juansc great test cases! Thanks for thinking through what our precedence/associativity should be for the `**` operator. Once we update the macrosyntax, this looks good to me.

File: README.md (hunk 1 above)
Comment (juansc) [inline README.md]: That looks right.

**Comment by rachelriv** [comment]
Make sure that you escape single quotes within the AST now.

**Comment by juansc** [comment]
All my tests failed when I merged with master. What's going on?

**Comment by alexschneider** [comment]
@juansc, #53 changed how errors are handled - rather than returning a map of the error and result, it just returns the result and throws an exception if we have an error.

File: test/parser/parser_spec.coffee [hunk 3]
```diff
@@ -51,6 +51,17 @@ describe 'Parser', ->
           expect(program.toString()).to.eql outputASTs.program9_AST_string
           done()
 
+    context 'when we have exponentsm they are right associative and
+              unary negation has different precedence on the left and right
+              of the ** operator', ->
+
+      it 'parses correctly', (done) ->
+        scan "#{validProgramsPath}/program14.tea", (err, tokens) ->
+          {error, program} = parse tokens
```
Comment (alexschneider) [inline test/parser/parser_spec.coffee]: This should just be `program = parse tokens`

File: test/parser/parser_spec.coffee [hunk 4]
```diff
@@ -51,6 +51,17 @@ describe 'Parser', ->
...
+          expect(error).to.be.null
```
Comment (alexschneider) [inline test/parser/parser_spec.coffee]: This line isn't needed anymore.

File: README.md [hunk 5]
```diff
@@ -68,7 +68,7 @@ Exp2    ::=  Exp3 (relop Exp3)?
 Exp3    ::=  Exp4 (addop Exp4)*
 Exp4    ::=  Exp5 (mulop Exp5)*
 Exp5    ::=  prefixop? Exp6
-Exp6    ::=  (Exp7 unary?) | (unary Exp7)
+Exp6    ::=  Exp7 ('**' Exp5)?
```
Comment (rachelriv) [inline README.md:71]: nice :+1:

**Comment by juansc** [comment]
@whusted @rachelriv @alexschneider Does this look fine to merge?

**Comment by rachelriv** [comment]
@juansc Looks good to me. :tea: :smile: :+1:

**Comment by whusted** [comment]
Me too

**Pull request merged**